### ISSUE TITLE
PDF refactor phase 2

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/firefox-ios#24661]
For phase 2 is expected:

* add test coverage for new implementation
* Cancel download when toolbar X button is pressed
* Correct the tab name to be the document name instead of `file://`
* Add missing Accessibility for `TemporaryDocumentLoadingView`

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-11331)

**Comment by data-sync-user** [comment]
➤ Filippo Zazzeroni commented:

For the PDF refactor here is the test cases that should be validated:

  1. The PDF loading view appears correctly when navigating to PDF
  2. The tab name for a PDF should have the PDF name instead of “file:// ( file://” )”
  3. if the app is closed when the app is reopened the PDF is still visible (If a PDF was viewed in the previous session)
  4. The PDF loading view is removed with the old and new toolbar when navigating back, forward, or when cancelling the load. You can test this with this big  PDF ( https://www.latexstudio.net/wp-content/uploads/2014/12/Game_Engine_Architecture-en.pdf )
  5. When a big PDF is loaded the sharing of it is fast, since it is already downloaded.
  6. Downloading a PDF from a website that needs login doesn’t download a broken PDF. i used this website for check https://online.emea.adp.com/signin/v1?APPID=ADPworld&productId=d3012000-a7b7-6167-e053-f282530b58ef&returnURL=https://adpworld.adp.com/&callingAppId=ADPworld&TARGET=-SM-https://adpworld.adp.com/ ( https://online.emea.adp.com/signin/v1?APPID=ADPworld&productId=d3012000-a7b7-6167-e053-f282530b58ef&returnURL=https://adpworld.adp.com/&callingAppId=ADPworld&TARGET=-SM-https://adpworld.adp.com/|smart-link ) . I don’t know if u are using the same websites for pay role, other wise we can do the test together.

Hi Alina Moldovan here is a list of all the tests for PDF refactor, let me know if something is not clear, Thanks :thanks: 

cc Andres Furlan

**Comment by oli-ver** [comment]
@FilippoZazzeroni Not sure if this is the right place to discuss this: The JIRA Issue mentioned in phase 1 cannot be reviewed without JIRA permission. Does this refactoring also include handling blobs properly? If a web page right now loads a blob PDF directly into the browser window object it is handled not properly in firefox as well, see discussion in #10690 and #14053

**Comment by FilippoZazzeroni** [comment]
Hi @oli-ver yes this should be correct right now, since we are downloading the PDF locally before displaying it to the web view. The web view doesn't download anymore the PDF so that part is skipped, this also fixes sharing, since previously some documents where viewed correctly by the webview but the sharing was broken due to missing cookies authorization when re-downloading the PDF before sharing.

Also you can check it if u want on the main branch, there all the fixes are present, but make sure to enable PDF refactor from the dev settings.

**Comment by oli-ver** [comment]
Thanks a lot, I will have a look at it.

**Comment by FilippoZazzeroni** [comment]
No problem @oli-ver please let us know if you find any bugs or have any suggestion. Consider tho the feature is still not stable.

**Comment by oli-ver** [comment]
@FilippoZazzeroni I tested this with the feature flag enabled. In the new implementation the PDF is not visualized anymore at all if it is loaded from a blob.

I tested using a small python reproduction application and one of the web pages where this happens for me. In both cases the PDF is now not visualized anymore:

<img width="355" alt="Image" src="https://github.com/user-attachments/assets/8b971d2a-951e-42a5-bde6-6a466538e113" />

In the logs I see that the download implementation seems not to support blob URLs:

```text
Task <50DD083B-34DE-40C4-9D7A-82CCE3823823>.<1> finished with error [-1002] Error Domain=NSURLErrorDomain Code=-1002 "unsupported URL" UserInfo={NSLocalizedDescription=unsupported URL, NSErrorFailingURLStringKey=blob:http://127.0.0.1:5000/2e232cc2-484c-49a1-92d7-1d1c2978fddf, NSErrorFailingURLKey=blob:http://127.0.0.1:5000/2e232cc2-484c-49a1-92d7-1d1c2978fddf, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDownloadTask <50DD083B-34DE-40C4-9D7A-82CCE3823823>.<1>"
), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDownloadTask <50DD083B-34DE-40C4-9D7A-82CCE3823823>.<1>, NSUnderlyingError=0x600000e4e820 {Error Domain=kCFErrorDomainCFNetwork Code=-1002 "(null)"}}
```
You can reproduce the behavior using the following example code:

```python
from flask import Flask, make_response
import urllib.request as request

app = Flask(__name__)

@app.route("/")
def hello_world():
    return  """
            <script>
                const getLocalPdfUrl = async () => {
                    const url = '/view_pdf';

                    const response = await fetch(url);
                    const blob = await response.blob();

                    const fileUrl = URL.createObjectURL(blob);
                    window.open(fileUrl, "_self")
                    console.log("Generated FileURL: " + fileUrl)
                    
                };
            </script>
            <a onclick="getLocalPdfUrl()">Open PDF</a>
            """ 


@app.route("/view_pdf")
def view_pdf():
    pdf_binary = read_pdf_binary()
    response = make_response(pdf_binary)
    response.headers['Content-Type'] = 'application/pdf'
    response.headers['ContentDisposition'] = \
            'attachment; filename="document.pdf"'
    return response


def read_pdf_binary():
    url = 'https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf'
    r = request.Request(url)
    f = request.urlopen(r)
    return f.read()
```
This is the requirements.txt
```text
cat requirements.txt 
click==8.1.8
Flask==2.2.5
importlib-metadata==6.7.0
itsdangerous==2.1.2
Jinja2==3.1.5
MarkupSafe==2.1.5
typing_extensions==4.7.1
Werkzeug==2.2.3
zipp==3.15.0
```

**Comment by oli-ver** [comment]
Great, thanks as well. Happy to help.

**Comment by FilippoZazzeroni** [comment]
i'll let you know as soon as this gets fixed 🎉  thanks again

**Comment by data-sync-user** [comment]
➤ Adina Petridean commented:

Verified as fixed on v137.0(52236) with iPhone 14 Pro Max (iOS 16.2).
